### ISSUE TITLE
TypeError: Cannot set property attributes of [object Object] which has only a getter

### ISSUE THREAD

**Issue by dmc1985** [atomiks/tippy.js-react#44]
I am in the process of writing tests using Enzyme for a React component that I created based on the Tippy component I imported from @tippy.js/react. When I called Enzyme's mount function using the Tooltip component I created with a nested DOM element, I receive the following error: TypeError: Cannot set property attributes of [object Object] which has only a getter.

Here is the code that causes the error: beforeEach(() => { wrapper = mount( <Tooltip content="content" theme="light" title="title" placement="bottom" > <div>Test</div> </Tooltip> ); })
However, when I replace <div>Test</div> with simple text, it works.

Any idea of how to resolve this issue? Thanks in advance for your help.

Here is the stack trace:
at polyfillElementPrototypeProperties (node_modules/tippy.js/dist/tippy.all.js:3379:27)
at tippy (node_modules/tippy.js/dist/tippy.all.js:4465:5)
at Function.Object..tippy.one (node_modules/tippy.js/dist/tippy.all.js:4504:10)
at Tippy.componentDidMount (node_modules/@tippy.js/react/dist/Tippy.js:141:26)
at commitLifeCycles (node_modules/react-dom/cjs/react-dom.development.js:14684:22)
at commitAllLifeCycles (node_modules/react-dom/cjs/react-dom.development.js:15904:7)
at HTMLUnknownElement.callCallback (node_modules/react-dom/cjs/react-dom.development.js:145:14)
at invokeEventListeners (node_modules/jsdom/lib/jsdom/living/events/EventTarget-impl.js:219:27)
at HTMLUnknownElementImpl._dispatch (node_modules/jsdom/lib/jsdom/living/events/EventTarget-impl.js:126:9)
at HTMLUnknownElementImpl.dispatchEvent (node_modules/jsdom/lib/jsdom/living/events/EventTarget-impl.js:87:17)
at HTMLUnknownElementImpl.dispatchEvent (node_modules/jsdom/lib/jsdom/living/nodes/HTMLElement-impl.js:36:27)
at HTMLUnknownElement.dispatchEvent (node_modules/jsdom/lib/jsdom/living/generated/EventTarget.js:61:35)
at Object.invokeGuardedCallbackDev (node_modules/react-dom/cjs/react-dom.development.js:195:16)
at invokeGuardedCallback (node_modules/react-dom/cjs/react-dom.development.js:248:31)
at commitRoot (node_modules/react-dom/cjs/react-dom.development.js:16074:7)
at completeRoot (node_modules/react-dom/cjs/react-dom.development.js:17462:3)
at performWorkOnRoot (node_modules/react-dom/cjs/react-dom.development.js:17390:9)
at performWork (node_modules/react-dom/cjs/react-dom.development.js:17294:7)
at performSyncWork (node_modules/react-dom/cjs/react-dom.development.js:17266:3)
at requestWork (node_modules/react-dom/cjs/react-dom.development.js:17154:5)
at scheduleWork (node_modules/react-dom/cjs/react-dom.development.js:16948:5)
at scheduleRootUpdate (node_modules/react-dom/cjs/react-dom.development.js:17636:3)
at updateContainerAtExpirationTime (node_modules/react-dom/cjs/react-dom.development.js:17663:10)
at updateContainer (node_modules/react-dom/cjs/react-dom.development.js:17690:10)
at ReactRoot.Object..ReactRoot.render (node_modules/react-dom/cjs/react-dom.development.js:17956:3)
at node_modules/react-dom/cjs/react-dom.development.js:18096:14
at unbatchedUpdates (node_modules/react-dom/cjs/react-dom.development.js:17517:10)
at legacyRenderSubtreeIntoContainer (node_modules/react-dom/cjs/react-dom.development.js:18092:5)
at Object.render (node_modules/react-dom/cjs/react-dom.development.js:18151:12)
at Object.render (node_modules/enzyme-adapter-react-16/build/ReactSixteenAdapter.js:314:114)
at new ReactWrapper (node_modules/enzyme/build/ReactWrapper.js:119:16)
at mount (node_modules/enzyme/build/mount.js:21:10)
at Object.test (src/components/Card/tests/Card.test.js:9:44)
at new Promise ()
at Promise.resolve.then.el (node_modules/p-map/index.js:46:16)
at 
at process._tickCallback (internal/process/next_tick.js:188:7)

**Comment by atomiks** [comment]
> polyfillElementPrototypeProperties

This means `tippy()` is receiving a plain object `{}` instead (maybe React element object?) of a DOM node

**Comment by dmc1985** [comment]
Thank you for your quick response. I see that when I use enzyme's mount, it returns a plain object which is causing the error per your explanation above. Is it possible to modify this plain object (such as adding a method) to prevent the error so that I may create working tests using the Tooltip component I created based on Tippy?

**Comment by atomiks** [comment]
Can you post the component here?

**Comment by dmc1985** [comment]
Here is the code for the component: 
`function Tooltip(props) {
    return (
        <Tippy
            arrow={false}
            allowHTML={true}
            animateFill={false}
            animation='shift-away'
            duration={300}
            maxWidth="130px"
            placement={props.placement}
            theme={props.theme}
            content={(
                <div data-brain='tooltip'>
                    <div className='tooltip-content'>
                        <p className='title'>{props.title}</p>
                        <p className='content'>{props.content}</p>
                    </div>
                </div>
            )}
        >
            {props.children}
        </Tippy>
    );
}`

And here is the code for the test that causes the problem: 
import { mount } from 'enzyme'; 
`wrapper = mount(
            <Tooltip
                content="content"
                theme="light"       
                title="title"
                placement="bottom"
            >
                <Button>Test</Button>
            </Tooltip>
        )`

**Comment by atomiks** [comment]
Hm can you show the code for `<Button>`?

**Comment by dmc1985** [comment]
Here is the code for the Button component: 
```
function Button(props) {
    const { children, type, disabled, ...otherProps } = props;
    const submit = type === 'classic' ? 'submit' : 'button';
    return (
        <button
            data-brain="button"
            className={disabled ? 'disable' : `button-${type}`}
            type={submit}
            disabled={disabled}
            {...otherProps}
        >
            {children}
        </button>
    );
}
```

However, I don't think this component is the source of the problem because if I put any sort of HTML element as the child of the Tooltip component in the argument for enzyme's element. I tested it with <div>Test</div> and I got the same error as the title of this issue. The tests only pass when I use plain text that is not placed inside html.

I also added the tooltip to a Card component, and when I run the tests on this Card component which involve using enzyme's mount (these tests were all passing before adding the Tooltip) and now these tests fail. Here's the snippet of my use of the Tooltip in this Card component: 

          ```
            <Tooltip
                content={props.title}
                placement='bottom'
                theme='dark'
            >
                <h4>{props.title}</h4>
            </Tooltip>
```

**Comment by atomiks** [comment]
Try editing `node_modules/@tippy.js/react/dist/Tippy.js` and add 

```jsx
console.log(ReactDOM.findDOMNode(this))
```

in cDM and run the tests to see what's being passed

**Comment by dmc1985** [comment]
Per your request, I added `console.log(ReactDOM.findDOMNode(this))` to the componentDidMount lifecycle method of node_modules/@tippy.js/react/dist/Tippy.js (line 138). Here is the console.log in context just to be sure we're on the same page:

```
_createClass(Tippy, [{
      key: "componentDidMount",
      value: function componentDidMount() {
        console.log('console test', ReactDOM.findDOMNode(this))
        this.setState({
          isMounted: true
        });
```

Here is what I see in my console:
```
HTMLDivElement {
      '__reactInternalInstance$b72ohub83fu': 
       FiberNode {
         tag: 7,
         key: null,
         type: 'div',
         stateNode: [Circular],
         return: 
          FiberNode {
            tag: 2,
            key: null,
            type: [Object],
            stateNode: [Object],
            return: [Object],
            child: [Circular],
            sibling: null,
            index: 0,
            ref: null,
            pendingProps: [Object],
            memoizedProps: [Object],
            updateQueue: null,
            memoizedState: [Object],
            firstContextDependency: null,
            mode: 0,
            effectTag: 5,
            nextEffect: [Object],
            firstEffect: null,
            lastEffect: null,
            expirationTime: 0,
            childExpirationTime: 0,
            alternate: null,
            actualDuration: 0,
            actualStartTime: -1,
            selfBaseDuration: 0,
            treeBaseDuration: 0,
            _debugID: 51,
            _debugSource: [Object],
            _debugOwner: [Object],
            _debugIsCurrentlyTiming: false },
         child: null,
         sibling: null,
         index: 0,
         ref: null,
         pendingProps: { children: 'Test' },
         memoizedProps: { children: 'Test' },
         updateQueue: null,
         memoizedState: null,
         firstContextDependency: null,
         mode: 0,
         effectTag: 0,
         nextEffect: null,
         firstEffect: null,
         lastEffect: null,
         expirationTime: 0,
         childExpirationTime: 0,
         alternate: null,
         actualDuration: 0,
         actualStartTime: -1,
         selfBaseDuration: 0,
         treeBaseDuration: 0,
         _debugID: 52,
         _debugSource: 
          { fileName: '/brain/brain-app/src/components/Tooltip/_tests_/Tooltip.test.js',
            lineNumber: 16 },
         _debugOwner: null,
         _debugIsCurrentlyTiming: false },
      '__reactEventHandlers$b72ohub83fu': { children: 'Test' },
      [Symbol(impl)]: 
       HTMLDivElementImpl {
         _eventListeners: {},
         _core: 
          { DOMException: [Object],
            NamedNodeMap: [Function: NamedNodeMap],
            Attr: [Function: Attr],
            Node: [Object],
            Element: [Function: Element],
            DocumentFragment: [Function: DocumentFragment],
            HTMLDocument: [Function: Document],
            Document: [Function: Document],
            XMLDocument: [Function: XMLDocument],
            CharacterData: [Function: CharacterData],
            Text: [Function: Text],
            CDATASection: [Function: CDATASection],
            ProcessingInstruction: [Function: ProcessingInstruction],
            Comment: [Function: Comment],
            DocumentType: [Function: DocumentType],
            DOMImplementation: [Function: DOMImplementation],
            Event: [Object],
            CustomEvent: [Function: CustomEvent],
            MessageEvent: [Function: MessageEvent],
            ErrorEvent: [Function: ErrorEvent],
            HashChangeEvent: [Function: HashChangeEvent],
            FocusEvent: [Function: FocusEvent],
            PopStateEvent: [Function: PopStateEvent],
            UIEvent: [Function: UIEvent],
            MouseEvent: [Function: MouseEvent],
            KeyboardEvent: [Object],
            TouchEvent: [Function: TouchEvent],
            ProgressEvent: [Function: ProgressEvent],
            EventTarget: [Function: EventTarget],
            Location: [Function: Location],
            History: [Function: History],
            Blob: [Function: Blob],
            File: [Function: File],
            FileList: [Function: FileList],
            DOMParser: [Function: DOMParser],
            FormData: [Function: FormData],
            HTMLElement: [Function: HTMLElement],
            HTMLAnchorElement: [Function: HTMLAnchorElement],
            HTMLAppletElement: [Function: HTMLAppletElement],
            HTMLAreaElement: [Function: HTMLAreaElement],
            HTMLAudioElement: [Function: HTMLAudioElement],
            HTMLBaseElement: [Function: HTMLBaseElement],
            HTMLBodyElement: [Function: HTMLBodyElement],
            HTMLBRElement: [Function: HTMLBRElement],
            HTMLButtonElement: [Function: HTMLButtonElement],
            HTMLCanvasElement: [Function: HTMLCanvasElement],
            HTMLDataElement: [Function: HTMLDataElement],
            HTMLDataListElement: [Function: HTMLDataListElement],
            HTMLDialogElement: [Function: HTMLDialogElement],
            HTMLDirectoryElement: [Function: HTMLDirectoryElement],
            HTMLDivElement: [Function: HTMLDivElement],
            HTMLDListElement: [Function: HTMLDListElement],
            HTMLEmbedElement: [Function: HTMLEmbedElement],
            HTMLFieldSetElement: [Function: HTMLFieldSetElement],
            HTMLFontElement: [Function: HTMLFontElement],
            HTMLFormElement: [Function: HTMLFormElement],
            HTMLFrameElement: [Function: HTMLFrameElement],
            HTMLFrameSetElement: [Function: HTMLFrameSetElement],
            HTMLHeadingElement: [Function: HTMLHeadingElement],
            HTMLHeadElement: [Function: HTMLHeadElement],
            HTMLHRElement: [Function: HTMLHRElement],
            HTMLHtmlElement: [Function: HTMLHtmlElement],
            HTMLIFrameElement: [Function: HTMLIFrameElement],
            HTMLImageElement: [Function: HTMLImageElement],
            HTMLInputElement: [Function: HTMLInputElement],
            HTMLLabelElement: [Function: HTMLLabelElement],
            HTMLLegendElement: [Function: HTMLLegendElement],
            HTMLLIElement: [Function: HTMLLIElement],
            HTMLLinkElement: [Function: HTMLLinkElement],
            HTMLMapElement: [Function: HTMLMapElement],
            HTMLMediaElement: [Object],
            HTMLMenuElement: [Function: HTMLMenuElement],
            HTMLMetaElement: [Function: HTMLMetaElement],
            HTMLMeterElement: [Function: HTMLMeterElement],
            HTMLModElement: [Function: HTMLModElement],
            HTMLObjectElement: [Function: HTMLObjectElement],
            HTMLOListElement: [Function: HTMLOListElement],
            HTMLOptGroupElement: [Function: HTMLOptGroupElement],
            HTMLOptionElement: [Function: HTMLOptionElement],
            HTMLOutputElement: [Function: HTMLOutputElement],
            HTMLParagraphElement: [Function: HTMLParagraphElement],
            HTMLParamElement: [Function: HTMLParamElement],
            HTMLPreElement: [Function: HTMLPreElement],
            HTMLProgressElement: [Function: HTMLProgressElement],
            HTMLQuoteElement: [Function: HTMLQuoteElement],
            HTMLScriptElement: [Function: HTMLScriptElement],
            HTMLSelectElement: [Function: HTMLSelectElement],
            HTMLSourceElement: [Function: HTMLSourceElement],
            HTMLSpanElement: [Function: HTMLSpanElement],
            HTMLStyleElement: [Function: HTMLStyleElement],
            HTMLTableCaptionElement: [Function: HTMLTableCaptionElement],
            HTMLTableCellElement: [Function: HTMLTableCellElement],
            HTMLTableColElement: [Function: HTMLTableColElement],
            HTMLTableDataCellElement: [Function: HTMLTableDataCellElement],
            HTMLTableElement: [Function: HTMLTableElement],
            HTMLTableHeaderCellElement: [Function: HTMLTableHeaderCellElement],
            HTMLTimeElement: [Function: HTMLTimeElement],
            HTMLTitleElement: [Function: HTMLTitleElement],
            HTMLTableRowElement: [Function: HTMLTableRowElement],
            HTMLTableSectionElement: [Function: HTMLTableSectionElement],
            HTMLTemplateElement: [Function: HTMLTemplateElement],
            HTMLTextAreaElement: [Function: HTMLTextAreaElement],
            HTMLTrackElement: [Object],
            HTMLUListElement: [Function: HTMLUListElement],
            HTMLUnknownElement: [Function: HTMLUnknownElement],
            HTMLVideoElement: [Function: HTMLVideoElement],
            StyleSheet: [Function: StyleSheet],
            MediaList: [Function: MediaList],
            CSSStyleSheet: [Function: CSSStyleSheet],
            CSSRule: [Object],
            CSSStyleRule: [Object],
            CSSMediaRule: [Function: CSSMediaRule],
            CSSImportRule: [Function: CSSImportRule],
            CSSStyleDeclaration: [Function: CSSStyleDeclaration],
            StyleSheetList: [Function: StyleSheetList],
            XPathException: [Object],
            XPathExpression: [Function: XPathExpression],
            XPathResult: [Object],
            XPathEvaluator: [Function: XPathEvaluator],
            HTMLCollection: [Function: HTMLCollection],
            NodeFilter: [Object],
            NodeIterator: [Function: NodeIterator],
            NodeList: [Function: NodeList],
            XMLHttpRequestEventTarget: [Function: XMLHttpRequestEventTarget],
            XMLHttpRequestUpload: [Function: XMLHttpRequestUpload],
            DOMTokenList: [Function: DOMTokenList],
            URL: [Function: URL],
            Window: [Function: Window] },
         _ownerDocument: 
          DocumentImpl {
            _eventListeners: {},
            _core: [Object],
            _ownerDocument: [Circular],
            _childNodesList: null,
            _childrenList: null,
            _version: 2,
            _memoizedQueries: {},
            nodeType: 9,
            _parsingMode: 'html',
            _htmlToDom: [Object],
            _implementation: [Object],
            _defaultView: [Object],
            _global: [Object],
            _documentElement: [Object],
            _ids: {},
            _attached: true,
            _currentScript: null,
            _cookieJar: [Object],
            _contentType: undefined,
            _encoding: 'UTF-8',
            _URL: [Object],
            _origin: 'http://localhost',
            _location: [Object],
            _history: [Object],
            _activeNodeIterators: [],
            _activeNodeIteratorsMax: 10,
            _referrer: '',
            _lastModified: '01/28/2019 11:12:59',
            _queue: [Object],
            _customResourceLoader: undefined,
            _pool: [Object],
            _agentOptions: [Object],
            _strictSSL: true,
            _proxy: undefined,
            _requestManager: [Object],
            readyState: 'complete',
            _lastFocusedElement: null,
            [Symbol(DOM SymbolTree)]: [Object],
            [Symbol(wrapper)]: [Object] },
         _childNodesList: null,
         _childrenList: null,
         _version: 2,
         _memoizedQueries: {},
         nodeType: 1,
         scrollTop: 0,
         scrollLeft: 0,
         _namespaceURI: 'http://www.w3.org/1999/xhtml',
         _prefix: null,
         _localName: 'div',
         _attributes: NamedNodeMap { [Symbol(NamedNodeMap internal)]: [Object] },
         _tabIndex: 0,
         _settingCssText: false,
         _clickInProgress: false,
         _style: 
          CSSStyleDeclaration {
            _values: {},
            _importants: {},
            _length: 0,
            _onChange: [Function] },
         [Symbol(DOM SymbolTree)]: 
          SymbolTreeNode {
            parent: [Object],
            previousSibling: null,
            nextSibling: null,
            firstChild: [Object],
            lastChild: [Object],
            childrenVersion: 1,
            childIndexCachedUpTo: null,
            cachedIndex: -1,
            cachedIndexVersion: NaN },
         [Symbol(wrapper)]: [Circular] } }
```

**Comment by atomiks** [comment]
Okay, that's strange I guess... I don't know why it's seeing it as a plain object (which uses the naive check `{}.hasOwnProperty.call(ReactDOM.findDOMNode(this)) === '[object Object]'` - maybe log that out?)

If all else fails, I'd suggest switching to `react-resting-library` instead of `enzyme` (that's what I switched to to test this library because enzyme was giving me other troubles)

**Comment by KubaJastrz** [comment]
I have been testing `Tippy` with enzyme (only `shallow` though) for a while now. I will post a working example later today if you want.

**Comment by atomiks** [comment]
I just set up a basic directory to test this out and didn't experience this error:

index.test.js

```jsx
import React from 'react'
import Tippy from '@tippy.js/react'
import { mount, configure } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'

configure({ adapter: new Adapter() })

function Tooltip(props) {
  return (
    <Tippy
      arrow={false}
      allowHTML={true}
      animateFill={false}
      animation="shift-away"
      duration={300}
      maxWidth="130px"
      placement={props.placement}
      theme={props.theme}
      content={
        <div data-brain="tooltip">
          <div className="tooltip-content">
            <p className="title">{props.title}</p>
            <p className="content">{props.content}</p>
          </div>
        </div>
      }
    >
      {props.children}
    </Tippy>
  )
}

test('<Tippy />', () => {
  const wrapper = mount(
    <Tooltip content="content" theme="light" title="title" placement="bottom">
      <h1>example</h1>
    </Tooltip>,
  )
  expect(true).toBe(true)
})
```

package.json

```json
{
  "name": "tippy-testing",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "scripts": {
    "test": "jest --coverage"
  },
  "author": "",
  "license": "ISC",
  "jest": {
    "setupFilesAfterEnv": [
      "<rootDir>setup.js"
    ],
    "coveragePathIgnorePatterns": [
      "test/setup.js"
    ]
  },
  "dependencies": {
    "@tippy.js/react": "^1.1.1",
    "jest": "^24.0.0",
    "react": "^16.7.0",
    "react-dom": "^16.7.0"
  },
  "devDependencies": {
    "@babel/core": "^7.2.2",
    "@babel/preset-env": "^7.3.1",
    "@babel/preset-react": "^7.0.0",
    "babel-jest": "^24.0.0",
    "enzyme": "^3.8.0",
    "enzyme-adapter-react-16": "^1.8.0"
  }
}
```

setup.js

```js
global.MutationObserver = class {
  constructor() {}
  disconnect() {}
  observe() {}
}

global.window.document.createRange = function createRange() {
  return {
    setEnd: () => {},
    setStart: () => {},
    getBoundingClientRect: () => {
      return { right: 0 }
    },
    getClientRects: () => [],
    commonAncestorContainer: document.createElement('div'),
  }
}

global.window.focus = () => {}
global.window.scroll = () => {}
```

**Comment by dmc1985** [comment]
@atomiks I tried to recreate your simple Tippy test but I got a lot of errors concerning babel and jest. Could you please upload the entire project to github so I could fork the repo? 

@KubaJastrz although @atomiks is right that shallow isn't exactly what I need, I'd be curious to see your working exmaple.

**Comment by KubaJastrz** [comment]
Everything should _just work_ -> https://github.com/KubaJastrz/tippy-testing

My general rule of thumb when using `enzyme`: don't use `mount` unless really have to (ie. using context which isn't supported 😢)

**Comment by wakame-isono** [comment]
Hello.
I hava the same error by Vue & Jest.
I found it: https://github.com/jsdom/jsdom/issues/766.
jsdom under 10.1.0 return [object Object] even if it is an Element by Object.prototype.toString.call.
They supported it after 10.1.0.
See: https://github.com/jsdom/jsdom/blob/16d3913eea0360c7757e75f7e266c3873c85b7dd/Changelog.md#1010

Thank you for the awesome package!
It is of great help to me.

**Comment by atomiks** [comment]
@wakame-isono can you setup a minimal reproduction repo for me? In the PR it looks like it should return the constructor in the string, like `[object HTMLDivElement]`, instead of `[object Object]`

**Comment by wakame-isono** [comment]
@atomiks 
I created repository for report.
https://github.com/wakame-isono/tippy-vue-testing

When yarn run test, {}.toString.call(Element) result is output to the console.

I think we should probably use jsdom over ver 10.1.0.
if someone jsdom less then ver 10.1.0, we can mock tippy.js.

for example.

Create moke file to `<rootDir>/__mocks__/tippy.js.js`

then

```
const tippy = jest.fn();

Object.defineProperty(tippy, 'setDefaults', {
  get() {
    return jest.fn();
  },
});

module.exports = tippy;
```

**Comment by atomiks** [comment]
We probably can do this instead:

```js
{}.hasOwnProperty.call(value, 'getBoundingClientRect')
```

A "virtual element" requires this property. Elements, NodeLists, strings, arrays, don't have this as an "own" property. Should work?

**Comment by atomiks** [comment]
This seems to work (tested it):

```js
export function isVirtualElement(value) {
  return (
    {}.toString.call(value) === '[object Object]' && !value.addEventListener
  )
}
```

Virtual objects passed to tippy won't have `addEventListener`, but jsdom objects do.

**Comment by wakame-isono** [comment]
@atomiks 
Sorry I am not sure React virtual DOM.
But it also work well for me!!  ✨ ✨ 👏👏

**Comment by atomiks** [comment]
A fix for this was released in 4.0.3 of the native library
